loadtesting causes stats.json to fail to load sometimes

### Description of the bug / feature
I ran some loadtests with apache jmeter. The application works fine during debugging and also during the loadtests until some threshold (around 1000 users). Above that threshold during UI Initialization the application fails to load the stats.json file (I guess due too many open files).
### Minimal reproducible example
Haven't had the time yet to create a minimal example.
### Expected behavior
Opening a ui should cause the server to open the same file each time.
### Actual behavior
It fails to open stats.json
### Versions:
    - Vaadin / Flow version: Vaadin 14.4.7 (production mode, compatibility mode disabled)
    - Java version: 8
    - OS version: debian 9
    - Browser version (if applicable): apache jmeter
    - Application Server (if applicable): tomcat 8
    - IDE (if applicable):

I tried to capture some useful messages from the logs:
```
ERROR [http-nio-8080-exec-71] (FrontendUtils:505) - Cannot get the 'stats.json' from the classpath 'META-INF/VAADIN/config/stats.json'
WARN  [http-nio-8080-exec-71] (UIInitListener:67) - Unexpected exception occured: BootstrapException: Unable to read webpack stats file.
DEBUG [http-nio-8080-exec-71] (UIInitListener:70) - Unexpected exception occured: com.vaadin.flow.server.BootstrapException: Unable to read webpack stats file.
        at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupFrameworkLibraries(BootstrapHandler.java:894)
        at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupDocumentHead(BootstrapHandler.java:771)
        at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.getBootstrapPage(BootstrapHandler.java:540)
        at com.vaadin.flow.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHandler.java:481)
        at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1547)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:247)
        at com.vaadin.cdi.CdiVaadinServlet.service(CdiVaadinServlet.java:67)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.openejb.server.httpd.EEFilter.doFilter(EEFilter.java:65)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:200)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.tomee.catalina.OpenEJBValve.invoke(OpenEJBValve.java:45)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:543)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.tomee.catalina.OpenEJBSecurityListener$RequestCapturer.invoke(OpenEJBSecurityListener.java:97)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:615)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1627)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: The stats file from webpack (stats.json) was not found.
The application is running in production mode.Verify that build-frontend task has executed successfully and that stats.json is on the classpath.Or switch application to development mode.
        at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.appendNpmBundle(BootstrapHandler.java:926)
        at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupFrameworkLibraries(BootstrapHandler.java:892)
        ... 38 more


```

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -595,6 +595,7 @@ public class FrontendUtils {
                         .toLocalDateTime();
                 Stats statistics = context.getAttribute(Stats.class);
                 if (statistics == null
+                        || statistics.getLastModified() == null
                         || modified.isAfter(statistics.getLastModified())) {
                     statistics = new Stats(
                             streamToString(connection.getInputStream()),
@@ -627,6 +628,13 @@ public class FrontendUtils {
     }
 
     private static InputStream getStatsFromClassPath(VaadinService service) {
+        Stats statistics = service.getContext().getAttribute(Stats.class);
+
+        if (statistics != null) {
+            return new ByteArrayInputStream(
+              statistics.statsJson.getBytes(StandardCharsets.UTF_8));
+        }
+
         String stats = service.getDeploymentConfiguration()
                 .getStringProperty(SERVLET_PARAMETER_STATISTICS_JSON,
                         VAADIN_SERVLET_RESOURCES + STATISTICS_JSON_DEFAULT)
@@ -638,6 +646,12 @@ public class FrontendUtils {
         InputStream stream = null;
         try {
             stream = statsUrl == null ? null : statsUrl.openStream();
+            if (stream != null) {
+                statistics = new Stats(streamToString(stream), null);
+                service.getContext().setAttribute(statistics);
+                stream = new ByteArrayInputStream(
+                  statistics.statsJson.getBytes(StandardCharsets.UTF_8));
+            }
         } catch (IOException exception) {
             getLogger().warn("Couldn't read content of stats file {}", stats,
                     exception);
@@ -1040,6 +1054,9 @@ public class FrontendUtils {
          * @return timestamp as LocalDateTime
          */
         public LocalDateTime getLastModified() {
+            if (lastModified == null) {
+                return null;
+            }
             return ZonedDateTime
                     .parse(lastModified, DateTimeFormatter.RFC_1123_DATE_TIME)
                     .toLocalDateTime();

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -444,6 +444,10 @@ public class FrontendUtils {
     /**
      * Gets the content of the <code>stats.json</code> file produced by webpack.
      *
+     * Note: Caches the <code>stats.json</code> when external stats is enabled or
+     * <code>stats.json</code> is provided from the class path. To clear the
+     * cache use {@link #clearCachedStatsContent(VaadinService)}.
+     *
      * @param service
      *            the vaadin service.
      * @return the content of the file as a string, null if not found.
@@ -470,6 +474,16 @@ public class FrontendUtils {
         return content != null
                 ? IOUtils.toString(content, StandardCharsets.UTF_8)
                 : null;
+    }
+
+    /**
+     * Clears the <code>stats.json</code> cache within this {@link VaadinContext}.
+     *
+     * @param service
+     *            the vaadin service.
+     */
+    public static void clearCachedStatsContent(VaadinService service) {
+        service.getContext().removeAttribute(Stats.class);
     }
 
     /**
@@ -672,9 +686,13 @@ public class FrontendUtils {
     }
 
     /**
-     * Load the asset chunks from stats.json. We will only read the file until
-     * we have reached the assetsByChunkName json and return that as a json
-     * object string.
+     * Load the asset chunks from <code>stats.json</code>. We will only read the
+     * file until we have reached the assetsByChunkName json and return that as
+     * a json object string.
+     *
+     * Note: The <code>stats.json</code> is cached when external stats is enabled
+     * or <code>stats.json</code> is provided from the class path. To clear the
+     * cache use {@link #clearCachedStatsContent(VaadinService)}.
      *
      * @param service
      *            the Vaadin service.

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -609,8 +609,7 @@ public class FrontendUtils {
                         .toLocalDateTime();
                 Stats statistics = context.getAttribute(Stats.class);
                 if (statistics == null
-                        || statistics.getLastModified() == null
-                        || modified.isAfter(statistics.getLastModified())) {
+                  || modified.isAfter(statistics.getLastModified().orElse(LocalDateTime.MIN))) {
                     statistics = new Stats(
                             streamToString(connection.getInputStream()),
                             lastModified);
@@ -1071,13 +1070,13 @@ public class FrontendUtils {
          *
          * @return timestamp as LocalDateTime
          */
-        public LocalDateTime getLastModified() {
+        public Optional<LocalDateTime> getLastModified() {
             if (lastModified == null) {
-                return null;
+                return Optional.empty();
             }
-            return ZonedDateTime
+            return Optional.of(ZonedDateTime
                     .parse(lastModified, DateTimeFormatter.RFC_1123_DATE_TIME)
-                    .toLocalDateTime();
+                    .toLocalDateTime());
         }
     }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -64,7 +64,7 @@ public class FrontendUtilsTest {
         try {
             CACHE_KEY = Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats");
         } catch (ClassNotFoundException e) {
-            System.err.println("Could not access cache key for stats.json!");
+            Assert.fail("Could not access cache key for stats.json!");
         }
     }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -51,6 +51,7 @@ import static com.vaadin.flow.server.Constants.VAADIN_SERVLET_RESOURCES;
 import static com.vaadin.flow.server.InitParameters.SERVLET_PARAMETER_STATISTICS_JSON;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
 public class FrontendUtilsTest {
@@ -326,6 +327,11 @@ public class FrontendUtilsTest {
       throws IOException, ClassNotFoundException, ServiceException {
         VaadinService service = setupStatsAssetMocks("ValidStats.json");
 
+        assertNull("Stats cache should not be present", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
+
+        // Populates cache
         FrontendUtils.getStatsContent(service);
 
         assertNotNull("Stats cache should be created", service.getContext().getAttribute(
@@ -338,6 +344,11 @@ public class FrontendUtilsTest {
       throws IOException, ClassNotFoundException, ServiceException {
         VaadinService service = setupStatsAssetMocks("ValidStats.json");
 
+        assertNull("Stats cache should not be present", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
+
+        // Populates cache
         FrontendUtils.getStatsAssetsByChunkName(service);
 
         assertNotNull("Stats cache should be created", service.getContext().getAttribute(
@@ -345,6 +356,27 @@ public class FrontendUtilsTest {
         ));
     }
 
+    @Test
+    public void clearCachedStatsContent_clearsCache()
+      throws IOException, ClassNotFoundException, ServiceException {
+        VaadinService service = setupStatsAssetMocks("ValidStats.json");
+
+        assertNull("Stats cache should not be present", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
+        // Can be invoked without cache - throws no exception
+        FrontendUtils.clearCachedStatsContent(service);
+
+        // Populates cache
+        FrontendUtils.getStatsContent(service);
+
+        // Clears cache
+        FrontendUtils.clearCachedStatsContent(service);
+
+        assertNull("Stats cache should not be present", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
+    }
     private ResourceProvider mockResourceProvider(VaadinService service) {
         DeploymentConfiguration config = Mockito
                 .mock(DeploymentConfiguration.class);

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -58,6 +58,16 @@ public class FrontendUtilsTest {
 
     private static final String USER_HOME = "user.home";
 
+    private static Class<?> CACHE_KEY;
+
+    static {
+        try {
+            CACHE_KEY = Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats");
+        } catch (ClassNotFoundException e) {
+            System.err.println("Could not access cache key for stats.json!");
+        }
+    }
+
     @Rule
     public ExpectedException exception = ExpectedException.none();
 
@@ -324,46 +334,41 @@ public class FrontendUtilsTest {
 
     @Test
     public void getStatsContent_getStatsFromClassPath_populatesStatsCache()
-      throws IOException, ClassNotFoundException, ServiceException {
+      throws IOException, ServiceException {
         VaadinService service = setupStatsAssetMocks("ValidStats.json");
 
-        assertNull("Stats cache should not be present", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNull("Stats cache should not be present",
+          service.getContext().getAttribute(CACHE_KEY));
 
         // Populates cache
         FrontendUtils.getStatsContent(service);
 
-        assertNotNull("Stats cache should be created", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNotNull("Stats cache should be created",
+          service.getContext().getAttribute(CACHE_KEY));
     }
 
     @Test
     public void getStatsAssetsByChunkName_getStatsFromClassPath_populatesStatsCache()
-      throws IOException, ClassNotFoundException, ServiceException {
+      throws IOException, ServiceException {
         VaadinService service = setupStatsAssetMocks("ValidStats.json");
 
-        assertNull("Stats cache should not be present", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNull("Stats cache should not be present",
+          service.getContext().getAttribute(CACHE_KEY));
 
         // Populates cache
         FrontendUtils.getStatsAssetsByChunkName(service);
 
-        assertNotNull("Stats cache should be created", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNotNull("Stats cache should be created",
+          service.getContext().getAttribute(CACHE_KEY));
     }
 
     @Test
     public void clearCachedStatsContent_clearsCache()
-      throws IOException, ClassNotFoundException, ServiceException {
+      throws IOException, ServiceException {
         VaadinService service = setupStatsAssetMocks("ValidStats.json");
 
-        assertNull("Stats cache should not be present", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNull("Stats cache should not be present",
+          service.getContext().getAttribute(CACHE_KEY));
         // Can be invoked without cache - throws no exception
         FrontendUtils.clearCachedStatsContent(service);
 
@@ -373,10 +378,10 @@ public class FrontendUtilsTest {
         // Clears cache
         FrontendUtils.clearCachedStatsContent(service);
 
-        assertNull("Stats cache should not be present", service.getContext().getAttribute(
-          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
-        ));
+        assertNull("Stats cache should not be present",
+          service.getContext().getAttribute(CACHE_KEY));
     }
+
     private ResourceProvider mockResourceProvider(VaadinService service) {
         DeploymentConfiguration config = Mockito
                 .mock(DeploymentConfiguration.class);

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -50,6 +50,7 @@ import static com.vaadin.flow.server.Constants.STATISTICS_JSON_DEFAULT;
 import static com.vaadin.flow.server.Constants.VAADIN_SERVLET_RESOURCES;
 import static com.vaadin.flow.server.InitParameters.SERVLET_PARAMETER_STATISTICS_JSON;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 
 public class FrontendUtilsTest {
@@ -292,6 +293,7 @@ public class FrontendUtilsTest {
                 manifestPaths.contains("/index.html"));
     }
 
+    @Test
     public void getStatsContent_getStatsFromClassPath_delegateToGetApplicationResource()
             throws IOException {
         VaadinServletService service = mockServletService();
@@ -317,6 +319,30 @@ public class FrontendUtilsTest {
         VaadinServlet servlet = service.getServlet();
 
         Mockito.verify(provider).getApplicationResource("foo");
+    }
+
+    @Test
+    public void getStatsContent_getStatsFromClassPath_populatesStatsCache()
+      throws IOException, ClassNotFoundException, ServiceException {
+        VaadinService service = setupStatsAssetMocks("ValidStats.json");
+
+        FrontendUtils.getStatsContent(service);
+
+        assertNotNull("Stats cache should be created", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
+    }
+
+    @Test
+    public void getStatsAssetsByChunkName_getStatsFromClassPath_populatesStatsCache()
+      throws IOException, ClassNotFoundException, ServiceException {
+        VaadinService service = setupStatsAssetMocks("ValidStats.json");
+
+        FrontendUtils.getStatsAssetsByChunkName(service);
+
+        assertNotNull("Stats cache should be created", service.getContext().getAttribute(
+          Class.forName("com.vaadin.flow.server.frontend.FrontendUtils$Stats")
+        ));
     }
 
     private ResourceProvider mockResourceProvider(VaadinService service) {

